src/crimson/osd/pg.cc

```diff
@@ -743,37 +745,58 @@ PG::do_osd_ops(
           std::move(txn),
           std::move(osd_op_p));
       });
-  }).safe_then_interruptible_tuple([this, m, obc, rvec = op_info.allows_returnvec()]()
-    -> PG::do_osd_ops_iertr::future<Ref<MOSDOpReply>> {
-    // TODO: should stop at the first op which returns a negative retval,
-    //       cmpext uses it for returning the index of first unmatched byte
-    int result = m->ops.empty() ? 0 : m->ops.back().rval.code;
-    if (result > 0 && !rvec) {
-      result = 0;
-    }
-    auto reply = make_message<MOSDOpReply>(m.get(),
-                                           result,
-                                           get_osdmap_epoch(),
-                                           0,
-                                           false);
-    reply->add_flags(CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK);
-    logger().debug(
-      "do_osd_ops: {} - object {} sending reply",
-      *m,
-      obc->obs.oi.soid);
-    return PG::do_osd_ops_ertr::make_ready_future<Ref<MOSDOpReply>>(
-      std::move(reply));
+  }).safe_then_unpack_interruptible_tuple(
+    [this, m, obc, rvec = op_info.allows_returnvec(), ox = ox.get()]
+    (auto submitted_fut, auto all_completed_fut)
+    -> PG::do_osd_ops_iertr::future<rep_op_fut_t> {
+    return seastar::make_ready_future<rep_op_fut_t>(
+          std::move(submitted_fut),
+          all_completed_fut.safe_then_interruptible_tuple(
+          [this, m, obc, rvec]()
+          -> PG::do_osd_ops_iertr::future<Ref<MOSDOpReply>> {
+            // TODO: should stop at the first op which returns a negative retval,
+            //       cmpext uses it for returning the index of first unmatched byte
+            int result = m->ops.empty() ? 0 : m->ops.back().rval.code;
+            if (result > 0 && !rvec) {
+              result = 0;
+            }
+            auto reply = make_message<MOSDOpReply>(m.get(),
+                                                   result,
+                                                   get_osdmap_epoch(),
+                                                   0,
+                                                   false);
+            reply->add_flags(CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK);
+            logger().debug(
+              "do_osd_ops: {} - object {} sending reply",
+              *m,
+              obc->obs.oi.soid);
+            return seastar::make_ready_future<Ref<MOSDOpReply>>(
+                std::move(reply));
+          }, crimson::ct_error::object_corrupted::handle([m, obc, this] {
+            return repair_object(m, obc->obs.oi.soid, obc->obs.oi.version)
```

Comment:
why do we need two `repair_object()`s:

1. one for the failure handling of `all_completed_fut`.
2. another for the failure handling of the outcome of `std::move(*ox).flush_changes(...)` call.